In [50]:
import pandas as pd
import numpy as np



/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [51]:
data= pd.read_excel("/content/Online retail.xlsx", header=None, names=['Items'])
data

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Items
0     shrimp,almonds,avocado,vegetables mix,green gr...
1                                burgers,meatballs,eggs
2                                               chutney
3                                        turkey,avocado
4     mineral water,milk,energy bar,whole wheat rice...
...                                                 ...
7496                      butter,light mayo,fresh bread
7497  burgers,frozen vegetables,eggs,french fries,ma...
7498                                            chicken
7499                                 escalope,green tea
7500    eggs,frozen smoothie,yogurt cake,low fat yogurt

[7501 rows x 1 columns]

In [52]:
data.isnull().sum()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Items    0
dtype: int64

In [53]:
data.duplicated()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0       False
1       False
2       False
3       False
4       False
        ...  
7496    False
7497    False
7498     True
7499     True
7500    False
Length: 7501, dtype: bool

In [54]:
data.duplicated().sum()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


2325

In [55]:
data.drop_duplicates(inplace=True)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [56]:
data.shape

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


(5176, 1)

In [57]:
data

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Items
0     shrimp,almonds,avocado,vegetables mix,green gr...
1                                burgers,meatballs,eggs
2                                               chutney
3                                        turkey,avocado
4     mineral water,milk,energy bar,whole wheat rice...
...                                                 ...
7493  burgers,salmon,pancakes,french fries,frozen sm...
7494  turkey,burgers,dessert wine,shrimp,pasta,tomat...
7496                      butter,light mayo,fresh bread
7497  burgers,frozen vegetables,eggs,french fries,ma...
7500    eggs,frozen smoothie,yogurt cake,low fat yogurt

[5176 rows x 1 columns]

In [58]:
data.reset_index(inplace=True)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [59]:
data

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


index                                              Items
0         0  shrimp,almonds,avocado,vegetables mix,green gr...
1         1                             burgers,meatballs,eggs
2         2                                            chutney
3         3                                     turkey,avocado
4         4  mineral water,milk,energy bar,whole wheat rice...
...     ...                                                ...
5171   7493  burgers,salmon,pancakes,french fries,frozen sm...
5172   7494  turkey,burgers,dessert wine,shrimp,pasta,tomat...
5173   7496                      butter,light mayo,fresh bread
5174   7497  burgers,frozen vegetables,eggs,french fries,ma...
5175   7500    eggs,frozen smoothie,yogurt cake,low fat yogurt

[5176 rows x 2 columns]

In [60]:
data= data['Items'].str.split(',', expand=True)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [61]:
data

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0                  1             2                 3   \
0            shrimp            almonds       avocado    vegetables mix   
1           burgers          meatballs          eggs              None   
2           chutney               None          None              None   
3            turkey            avocado          None              None   
4     mineral water               milk    energy bar  whole wheat rice   
...             ...                ...           ...               ...   
5171        burgers             salmon      pancakes      french fries   
5172         turkey            burgers  dessert wine            shrimp   
5173         butter         light mayo   fresh bread              None   
5174        burgers  frozen vegetables          eggs      french fries   
5175           eggs    frozen smoothie   yogurt cake    low fat yogurt   

                   4                 5       6               7             8   \
0        green grapes  whole weat flour    yams  cottage cheese  energy drink   
1                None              None    None            None          None   
2                None              None    None            None          None   
3                None              None    None            None          None   
4           green tea              None    None            None          None   
...               ...               ...     ...             ...           ...   
5171  frozen smoothie       fresh bread    mint            None          None   
5172            pasta          tomatoes  pepper            milk      pancakes   
5173             None              None    None            None          None   
5174        magazines         green tea    None            None          None   
5175             None              None    None            None          None   

                    9               10               11     12     13  \
0         tomato juice  low fat yogurt        green tea  honey  salad   
1                 None            None             None   None   None   
2                 None            None             None   None   None   
3                 None            None             None   None   None   
4                 None            None             None   None   None   
...                ...             ...              ...    ...    ...   
5171              None            None             None   None   None   
5172  whole wheat rice             oil  frozen smoothie   None   None   
5173              None            None             None   None   None   
5174              None            None             None   None   None   
5175              None            None             None   None   None   

                 14      15                 16               17       18  \
0     mineral water  salmon  antioxydant juice  frozen smoothie  spinach   
1              None    None               None             None     None   
2              None    None               None             None     None   
3              None    None               None             None     None   
4              None    None               None             None     None   
...             ...     ...                ...              ...      ...   
5171           None    None               None             None     None   
5172           None    None               None             None     None   
5173           None    None               None             None     None   
5174           None    None               None             None     None   
5175           None    None               None             None     None   

             19  
0     olive oil  
1          None  
2          None  
3          None  
4          None  
...         ...  
5171       None  
5172       None  
5173       None  
5174       None  
5175       None  

[5176 rows x 20 columns]

In [62]:
#find unique items from the dataset
items = set()
for col in data:
    items.update(data[col].unique())
print("Items count is ", len(items))
print(items)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Items count is  121
{'bug spray', 'dessert wine', 'fromage blanc', 'napkins', 'vegetables mix', 'fresh tuna', 'sparkling water', 'tomatoes', 'barbecue sauce', 'soda', 'tomato sauce', 'eggs', 'rice', 'strawberries', 'mushroom cream sauce', None, 'whole wheat pasta', 'cake', 'french wine', 'green grapes', 'pet food', 'green tea', 'honey', 'yogurt cake', 'salt', 'champagne', 'magazines', 'cookies', 'soup', 'mint green tea', 'muffins', 'salmon', 'bacon', ' asparagus', 'babies food', 'low fat yogurt', 'gluten free bar', 'toothpaste', 'mint', 'cooking oil', 'flax seed', 'cauliflower', 'chili', 'pepper', 'ham', 'mayonnaise', 'energy drink', 'parmesan cheese', 'yams', 'gums', 'spaghetti', 'frozen smoothie', 'water spray', 'red wine', 'whole wheat rice', 'mineral water', 'nonfat milk', 'olive oil', 'light cream', 'blueberries', 'pasta', 'burgers', 'turkey', 'herb & pepper', 'carrots', 'clothes accessories', 'pickles', 'avocado', 'strong cheese', 'shrimp', 'energy bar', 'light mayo', 'melons', '

In [63]:
from mlxtend.frequent_patterns import apriori,association_rules
from mlxtend.preprocessing import TransactionEncoder

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [64]:
data.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs              None          None   
2        chutney       None        None              None          None   
3         turkey    avocado        None              None          None   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1              None  None            None          None          None   
2              None  None            None          None          None   
3              None  None            None          None          None   
4              None  None            None          None          None   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1            None       None   None   None           None    None   
2            None       None   None   None           None    None   
3            None       None   None   None           None    None   
4            None       None   None   None           None    None   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1               None             None     None       None  
2               None             None     None       None  
3               None             None     None       None  
4               None             None     None       None

In [65]:
data.dtypes

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0     object
1     object
2     object
3     object
4     object
5     object
6     object
7     object
8     object
9     object
10    object
11    object
12    object
13    object
14    object
15    object
16    object
17    object
18    object
19    object
dtype: object

In [66]:
df_out = data.apply(lambda x: list(x.dropna().values), axis=1).tolist()

Encoder = TransactionEncoder()
dataset = Encoder.fit(df_out).transform(df_out)
dataset = dataset.astype('int')
print(dataset)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


[[0 1 1 ... 1 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 1 0]]


In [68]:
df = pd.DataFrame(dataset,columns= Encoder.columns_)
df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0              0        1                  1          0        1            0   
1              0        0                  0          0        0            0   
2              0        0                  0          0        0            0   
3              0        0                  0          0        1            0   
4              0        0                  0          0        0            0   
...          ...      ...                ...        ...      ...          ...   
5171           0        0                  0          0        0            0   
5172           0        0                  0          0        0            0   
5173           0        0                  0          0        0            0   
5174           0        0                  0          0        0            0   
5175           0        0                  0          0        0            0   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0         0               0          0            0  ...       0   
1         0               0          0            0  ...       0   
2         0               0          0            0  ...       0   
3         0               0          0            0  ...       1   
4         0               0          0            0  ...       0   
...     ...             ...        ...          ...  ...     ...   
5171      0               0          0            0  ...       0   
5172      0               0          0            0  ...       1   
5173      0               0          0            0  ...       0   
5174      0               0          0            0  ...       0   
5175      0               0          0            0  ...       0   

      vegetables mix  water spray  white wine  whole weat flour  \
0                  1            0           0                 1   
1                  0            0           0                 0   
2                  0            0           0                 0   
3                  0            0           0                 0   
4                  0            0           0                 0   
...              ...          ...         ...               ...   
5171               0            0           0                 0   
5172               0            0           0                 0   
5173               0            0           0                 0   
5174               0            0           0                 0   
5175               0            0           0                 0   

      whole wheat pasta  whole wheat rice  yams  yogurt cake  zucchini  
0                     0                 0     1            0         0  
1                     0                 0     0            0         0  
2                     0                 0     0            0         0  
3                     0                 0     0            0         0  
4                     0                 1     0            0         0  
...                 ...               ...   ...          ...       ...  
5171                  0                 0     0            0         0  
5172                  0                 1     0            0         0  
5173                  0                 0     0            0         0  
5174                  0                 0     0            0         0  
5175                  0                 0     0            1         0  

[5176 rows x 120 columns]

In [77]:
for col in df.columns:
    print(col)

 asparagus
almonds
antioxydant juice
asparagus
avocado
babies food
bacon
barbecue sauce
black tea
blueberries
body spray
bramble
brownies
bug spray
burger sauce
burgers
butter
cake
candy bars
carrots
cauliflower
cereals
champagne
chicken
chili
chocolate
chocolate bread
chutney
cider
clothes accessories
cookies
cooking oil
corn
cottage cheese
cream
dessert wine
eggplant
eggs
energy bar
energy drink
escalope
extra dark chocolate
flax seed
french fries
french wine
fresh bread
fresh tuna
fromage blanc
frozen smoothie
frozen vegetables
gluten free bar
grated cheese
green beans
green grapes
green tea
ground beef
gums
ham
hand protein bar
herb & pepper
honey
hot dogs
ketchup
light cream
light mayo
low fat yogurt
magazines
mashed potato
mayonnaise
meatballs
melons
milk
mineral water
mint
mint green tea
muffins
mushroom cream sauce
napkins
nonfat milk
oatmeal
oil
olive oil
pancakes
parmesan cheese
pasta
pepper
pet food
pickles
protein bar
red wine
rice
salad
salmon
salt
sandwich
shallot
shampoo

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [81]:
frequent_itemsets = apriori(df, min_support=0.05, use_colnames=True)
frequent_itemsets.sort_values(by=['support'],ascending=False)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                            itemsets
20  0.299845                     (mineral water)
26  0.229521                         (spaghetti)
7   0.208076                              (eggs)
4   0.205178                         (chocolate)
9   0.192620                      (french fries)
19  0.170015                              (milk)
14  0.169243                         (green tea)
15  0.135819                       (ground beef)
12  0.129830                 (frozen vegetables)
22  0.125193                          (pancakes)
0   0.113794                           (burgers)
1   0.103555                              (cake)
24  0.098918                            (shrimp)
18  0.095054                    (low fat yogurt)
27  0.091963                          (tomatoes)
21  0.087713                         (olive oil)
39  0.085008          (spaghetti, mineral water)
11  0.084815                   (frozen smoothie)
8   0.083849                          (escalope)
3   0.083849                           (chicken)
28  0.083269                            (turkey)
29  0.077859                  (whole wheat rice)
30  0.073223          (chocolate, mineral water)
6   0.071870                       (cooking oil)
25  0.070904                              (soup)
13  0.070131                     (grated cheese)
32  0.070131               (mineral water, eggs)
37  0.067813               (milk, mineral water)
16  0.066461                     (herb & pepper)
17  0.065301                             (honey)
5   0.060665                           (cookies)
10  0.059312                       (fresh bread)
35  0.058733        (mineral water, ground beef)
23  0.057573                            (salmon)
31  0.055835              (spaghetti, chocolate)
36  0.055835            (spaghetti, ground beef)
2   0.054869                         (champagne)
33  0.051391                   (spaghetti, eggs)
34  0.050425  (frozen vegetables, mineral water)
38  0.050039                   (spaghetti, milk)

In [82]:
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents          consequents  antecedent support  \
0           (chocolate)      (mineral water)            0.205178   
1       (mineral water)          (chocolate)            0.299845   
2           (spaghetti)          (chocolate)            0.229521   
3           (chocolate)          (spaghetti)            0.205178   
4       (mineral water)               (eggs)            0.299845   
5                (eggs)      (mineral water)            0.208076   
6           (spaghetti)               (eggs)            0.229521   
7                (eggs)          (spaghetti)            0.208076   
8   (frozen vegetables)      (mineral water)            0.129830   
9       (mineral water)  (frozen vegetables)            0.299845   
10      (mineral water)        (ground beef)            0.299845   
11        (ground beef)      (mineral water)            0.135819   
12          (spaghetti)        (ground beef)            0.229521   
13        (ground beef)          (spaghetti)            0.135819   
14               (milk)      (mineral water)            0.170015   
15      (mineral water)               (milk)            0.299845   
16          (spaghetti)               (milk)            0.229521   
17               (milk)          (spaghetti)            0.170015   
18          (spaghetti)      (mineral water)            0.229521   
19      (mineral water)          (spaghetti)            0.299845   

    consequent support   support  confidence      lift  leverage  conviction  \
0             0.299845  0.073223    0.356874  1.190193  0.011701    1.088674   
1             0.205178  0.073223    0.244201  1.190193  0.011701    1.051632   
2             0.205178  0.055835    0.243266  1.185635  0.008742    1.050332   
3             0.229521  0.055835    0.272128  1.185635  0.008742    1.058537   
4             0.208076  0.070131    0.233892  1.124070  0.007741    1.033698   
5             0.299845  0.070131    0.337047  1.124070  0.007741    1.056115   
6             0.208076  0.051391    0.223906  1.076078  0.003633    1.020397   
7             0.229521  0.051391    0.246982  1.076078  0.003633    1.023189   
8             0.299845  0.050425    0.388393  1.295310  0.011496    1.144778   
9             0.129830  0.050425    0.168170  1.295310  0.011496    1.046091   
10            0.135819  0.058733    0.195876  1.442184  0.018008    1.074686   
11            0.299845  0.058733    0.432432  1.442184  0.018008    1.233606   
12            0.135819  0.055835    0.243266  1.791102  0.024661    1.141988   
13            0.229521  0.055835    0.411095  1.791102  0.024661    1.308326   
14            0.299845  0.067813    0.398864  1.330231  0.016835    1.164718   
15            0.170015  0.067813    0.226160  1.330231  0.016835    1.072553   
16            0.170015  0.050039    0.218013  1.282316  0.011017    1.061380   
17            0.229521  0.050039    0.294318  1.282316  0.011017    1.091822   
18            0.299845  0.085008    0.370370  1.235204  0.016187    1.112010   
19            0.229521  0.085008    0.283505  1.235204  0.016187    1.075345   

    zhangs_metric  
0        0.201051  
1        0.228235  
2        0.203212  
3        0.196988  
4        0.157645  
5        0.139377  
6        0.091760  
7        0.089275  
8        0.262000  
9        0.325620  
10       0.437914  
11       0.354795  
12       0.573260  
13       0.511102  
14       0.299103  
15       0.354566  
16       0.285745  
17       0.265259  
18       0.247141  
19       0.271965

We can interpret the result of the first rule as:

The support is 0.07 calculated by dividing the number of transactions containing mineral water and chocolate by the total number of transactions.

The confidence level is 0.35 shows that out of all the transactions that contain chocolate, 0.35 transactions contain mineral water too.

The lift 1.19 tells us that mineral water is 1.19 times more likely to be bought by the customer who also buys chocolate.

1. What is lift and why is it important in Association rules?

Answer: Lift is a measure of how much two items are associated with each other.
Lift(A→B)= Support(A∪B)/Support(A)x Support(B)

 A lift value greater than 1 indicates a positive correlation between two items, while a value less than 1 indicates a negative correlation.
2. What is support and Confidence. How do you calculate them?

Support is the measure of how much times an item set appear in the dataset.

Support(A→B)= Number of transactions containing (A∪B)/Total number of transactions



Confidence is the measure of how likely two items appear together. It is the likelihood for item A to be purchased given item B is already purchased.

Confidence(A→B)= Support(A∪B)/Support(A)

​

3. What are some limitations or challenges of Association rules mining?

Limitations and Challenges of Association Rule Mining

Scalability:Mining association rules can be expensive, especially in large datasets, as the number of itemsets increases.

Redundancy:Many generated rules can be redundant or uninteresting. Removing insignificant rules can be challenging.

Threshold Sensitivity:The results of association rule mining heavily depend on the chosen support and confidence thresholds. Setting these thresholds too high may result in missing patterns, where as setting them too low may produce an large number of rules.

Actionability:Not all identified rules are actionable or useful in a practical context.

Dynamic Data:In environments where data is constantly changing, maintaining up-to-date association rules can be challenging and may require frequent recalculations.
